packages/protocol/lib/compatibility/verify-bytecode.ts

```diff
@@ -0,0 +1,277 @@
+  } else {
+    return getCoreContractAddress(contract, context)
+  }
+
+}
+
+const getOnchainBytecode = async (contract: string, isLibrary: boolean, context: VerificationContext): Promise<string> => {
+  const implementationAddress = await getImplementationAddress(contract, isLibrary, context)
+  const onchainBytecodeWithMetadata = await context.web3.eth.getCode(implementationAddress)
+  return stripMetadata(onchainBytecodeWithMetadata)
+}
+
+const dfsStep = async (queue: string[], visited: Set<string>, context: VerificationContext) => {
+  const contract = queue.pop()
+  if (ignoredContracts.includes(contract)) {
+    return
+  }
+
+  const isLibrary = !context.contracts.includes(contract)
+
+  const sourceBytecode = getSoureBytecode(contract, context)
+  const onchainBytecode = await getOnchainBytecode(contract, isLibrary, context)
+
+  const sourceLibraryPositions = collectLibraryPositions(sourceBytecode)
+  collectLibraryAddresses(onchainBytecode, sourceLibraryPositions, context.libraryAddresses)
+
+  const linkedSourceBytecode = linkLibraries(sourceBytecode, context.libraryAddresses)
+
+  // To check that a library isn't being called directly, the Solidity
+  // compiler starts a library's bytecode with a comparison of the current
+  // address with the address the library was deployed to (it has to differ
+  // to ensure the library is being called with CALLCODE or DELEGATECALL
+  // instead of a regular CALL).
+  // The address is only known at contract construction time, so
+  // the compiler's output contains a placeholder 0-address, while the onchain
+  // bytecode has the correct address inserted.
+  // Reference: https://solidity.readthedocs.io/en/v0.5.12/contracts.html#call-protection-for-libraries
+  let normalizedSourceBytecode
+  let normalizedOnchainBytecode
+  if (isLibrary) {
+    const implementationAddress = await getImplementationAddress(contract, true, context)
+
+    verifyLibraryPrefix(linkedSourceBytecode, '0000000000000000000000000000000000000000')
+    verifyLibraryPrefix(onchainBytecode, implementationAddress.slice(2, 42).toLowerCase())
+
+    normalizedSourceBytecode = linkedSourceBytecode.slice(44, linkedSourceBytecode.length)
+    normalizedOnchainBytecode = onchainBytecode.slice(44, onchainBytecode.length)
+  } else {
+    normalizedSourceBytecode = linkedSourceBytecode
+    normalizedOnchainBytecode = onchainBytecode
+  }
+
+  if (normalizedSourceBytecode !== normalizedOnchainBytecode) {
+    throw new Error(`${contract}'s onchain and compiled bytecodes do not match`)
+  }
+
+  Object.keys(sourceLibraryPositions).forEach(library => {
+    if (!visited.has(library)) {
+      queue.push(library)
+      visited.add(library)
+    }
+  })
+}
+
+export const verifyBytecodesDfs = async (contracts: string[], artifacts: BuildArtifacts, registry: RegistryInstance, proposal: ProposalTx[], Proxy: Truffle.Contract<ProxyInstance>, web3: Web3, isBeforeRelease1: boolean = false) => {
```

Comment:
Or conversely, this is arguably the meat of the functionality, it would be nice to have a high-level description of the approach of this verification as a comment to this function